addons/web/static/src/js/views/graph/graph_renderer.js

```diff
@@ -1,1079 +1,1049 @@
+                const labelCopy = label.slice(0);
+                if (originIndex !== undefined) {
+                    labelCopy.splice(i, 1, this.dateClasses.representative(label[i], originIndex));
+                } else {
+                    labelCopy.splice(i, 1, this.dateClasses.dateClassMembers(label[i]));
+                }
+                return labelCopy.join('/');
+            }
+            // here label is an array containing strings or numbers.
+            return label.join('/') || _t('Total');
         }
-        this.$el.empty();
-        if (!_.contains(CHART_TYPES, this.state.mode)) {
-            this.trigger_up('warning', {
-                title: _t('Invalid mode for chart'),
-                message: _t('Cannot render chart with mode : ') + this.state.mode
+
+        /**
+         * create bar chart.
+         *
+         * @private
+         * @param {Object[]} dataPoints
+         */
+        _renderBarChart(dataPoints) {
+            // prepare data
+            const data = this._prepareData(dataPoints);
+
+            for (let index = 0; index < data.datasets.length; ++index) {
+                const dataset = data.datasets[index];
+                // used when stacked
+                dataset.stack = this.props.stacked ? this.props.origins[dataset.originIndex] : undefined;
+                // set dataset color
+                var color = this._getColor(index);
+                dataset.backgroundColor = color;
+            }
+
+            // prepare options
+            const options = this._prepareOptions(data.datasets.length);
+
+            // create chart
+            const ctx = this.canvasRef.el;
+            this.chart = new Chart(ctx, {
+                type: 'bar',
+                data: data,
+                options: options,
             });
         }
-        var dataPoints = this._filterDataPoints();
-        if (!dataPoints.length && this.state.mode !== 'pie') {
-            this.$el.append(qweb.render('View.NoContentHelper'));
-        } else if (this.isInDOM) {
-            // only render the graph if the widget is already in the DOM (this
-            // happens typically after an update), otherwise, it will be
-            // rendered when the widget will be attached to the DOM (see
-            // 'on_attach_callback')
-            var $canvasContainer = $('<div/>', {class: 'o_graph_canvas_container'});
-            var $canvas = $('<canvas/>').attr('id', this.chartId);
-            $canvasContainer.append($canvas);
-            this.$el.append($canvasContainer);
-
-            var i = this.state.comparisonFieldIndex;
+
+        /**
+         * Render the chart if no message error is already displayed.
+         *
+         * @private
+         */
+        _renderChart() {
+            if (this.chart) {
+                this.chart.destroy();
+            }
+
+            if (this.noContentHelperData.show){
```

Comment:
I would prefer to put the test in mounted and patched. Clearer I think event if a bit longer